File: scripts/eval/phase3_lesson_lifecycle.py
"""Phase 3.5 Lesson Lifecycle Evaluation."""

import argparse
import json
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from longmem_dmr_smoke import eval_cargo_profile, repo_root


def run_phase3_lesson_lifecycle(output_path: Path, tag: str) -> dict:
    cargo_profile = eval_cargo_profile()
    cmd = ["cargo", "run"]
    if cargo_profile == "release":
        cmd.append("--release")
    cmd.extend(
        [
            "-p",
            "synapse-eval",
            "--bin",
            "kr-phase3-lesson-lifecycle",
            "--",
            "--json",
            str(output_path),
            "--tag",
            tag,
        ]
    )
    result = subprocess.run(cmd, cwd=repo_root(), text=True, capture_output=True)
    if result.returncode != 0:
        raise RuntimeError(
            "kr-phase3-lesson-lifecycle failed\nSTDOUT:\n{}\nSTDERR:\n{}".format(
                result.stdout, result.stderr
            )
        )
    print(result.stdout, end="")
    if result.stderr:
        print(result.stderr, end="", file=sys.stderr)
    report = json.loads(output_path.read_text(encoding="utf-8"))
    report["cargo_profile"] = cargo_profile
    return report


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run Phase 3.5 lesson lifecycle evaluation."
    )
    parser.add_argument(
        "--output-path",
        type=Path,
        default=repo_root() / "crates/eval/reports/phase3-lesson-lifecycle.json",
    )
    parser.add_argument("--tag", default="phase3-lesson-lifecycle")
    args = parser.parse_args()

    args.output_path.parent.mkdir(parents=True, exist_ok=True)
    report = run_phase3_lesson_lifecycle(args.output_path, args.tag)
    print(
        "summary: scenarios={} active={} challenged={} superseded={} candidate={} transition={:.4f} contradiction={:.4f} supersession={:.4f} reinforcement={:.4f} protection={:.4f} safety={:.4f} pass={}".format(
            report["scenarios"],
            report["states"]["active"],
            report["states"]["challenged"],
            report["states"]["superseded"],
            report["states"]["candidate"],
            report["metrics"]["lifecycle_transition_accuracy"],
            report["metrics"]["contradiction_response_score"],
            report["metrics"]["supersession_score"],
            report["metrics"]["reinforcement_score"],
            report["metrics"]["false_lesson_protection_score"],
            report["metrics"]["lifecycle_safety"],
            report["pass"],
        )
    )
    print(f"Saved to {args.output_path}")
    if report["pass"] and report["mode"] == "evaluation_only":
        print("Phase 3.5 Lesson Lifecycle Evaluation Completed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/eval/src/bin/phase3_lesson_lifecycle.rs
use anyhow::{Context, Result};
use clap::Parser;
use std::path::PathBuf;
use synapse_eval::Phase3LessonLifecycleEvaluator;

#[derive(Parser)]
#[command(
    name = "kr-phase3-lesson-lifecycle",
    about = "Synapse Phase 3.5 lesson lifecycle evaluation"
)]
struct Cli {
    /// Write the Phase3LessonLifecycleReport JSON to this path.
    #[arg(long)]
    json: PathBuf,
    /// Tag to embed in the report.
    #[arg(long, default_value = "phase3-lesson-lifecycle")]
    tag: String,
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    let report = Phase3LessonLifecycleEvaluator::evaluate(cli.tag)?;

    println!("=== Synapse Phase 3.5 lesson lifecycle evaluation ===");
    println!("tag:                              {}", report.tag);
    println!("phase:                            {}", report.phase);
    println!("mode:                             {}", report.mode);
    println!(
        "evaluation_version:               {}",
        report.evaluation_version
    );
    println!(
        "baseline_version:                 {}",
        report.baseline_version
    );
    println!("scenarios:                        {}", report.scenarios);
    println!(
        "future_influence_source_count:    {}",
        report.future_influence_source_count
    );
    println!(
        "state_candidate:                  {}",
        report.states.candidate
    );
    println!("state_active:                     {}", report.states.active);
    println!(
        "state_challenged:                 {}",
        report.states.challenged
    );
    println!(
        "state_superseded:                 {}",
        report.states.superseded
    );
    println!(
        "lifecycle_transition_accuracy:    {:.4}",
        report.metrics.lifecycle_transition_accuracy
    );
    println!(
        "contradiction_response_score:     {:.4}",
        report.metrics.contradiction_response_score
    );
    println!(
        "supersession_score:               {:.4}",
        report.metrics.supersession_score
    );
    println!(
        "reinforcement_score:              {:.4}",
        report.metrics.reinforcement_score
    );
    println!(
        "false_lesson_protection_score:    {:.4}",
        report.metrics.false_lesson_protection_score
    );
    println!(
        "lifecycle_safety:                 {:.4}",
        report.metrics.lifecycle_safety
    );
    println!(
        "memory_written:                   {}",
        report.safety.memory_written
    );
    println!(
        "playbook_created:                 {}",
        report.safety.playbook_created
    );
    println!(
        "future_influence_changed:         {}",
        report.safety.future_influence_changed
    );
    println!("pass:                             {}", report.pass);

    if let Some(parent) = cli.json.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&cli.json, serde_json::to_string_pretty(&report)?)
        .with_context(|| format!("writing JSON to {}", cli.json.display()))?;
    eprintln!(
        "wrote Phase 3.5 lesson lifecycle report to {}",
        cli.json.display()
    );
    Ok(())
}
